text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+
+### Validation
+
+The validator typechecks functions at use-sites via inlining.
+This means that functions can be polymorphic. For example, one could write the following, where `eq` is used in the first instance with a pair of entities, and in the second instance with a pair of strings.
+```
+function eq(a, b) {
+  a == b
+};
+
+permit(principal, action, resource)
+when {
+  eq(principal,resource.owner) ||
+  eq(principal.org,"Admin")
+};
+```
+
+## Drawbacks
+
+### Redundancy
+Cedar functions can only accomplish things that can already accomplished with Cedar expressions.
+This means we are not expanding the expressive power of Cedar in any way. 
+We are also adding more than one way to accomplish the same task.
+Bringing back our SemVer example:
+```
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+    resource.apiVersion.major > 2 ||
+    (resource.apiVersion.major == 2 && resource.apiVersion.minor >= 1)
+};
+```
+This policy does accomplish the user's goal of encoding the SemVer relationship.
+The problem is readability. A reader of this policy has to work out that it is implementing the standard semantic version comparison operator, and not some bespoke versioning scheme. Another problem is maintainability. If you have multiple policies that reason about SemVers, you have to repeat the logic, inline, in each policy. This violates basic software engineering tenets (Don’t Repeat Yourself), allowing bugs to sneak in via typos or botched copy/paste.
+
+### No custom parsing or error-handling
+Extension functions provide more full-featured constructors through custom parsing and error handling, but Cedar functions provide no such facilities. This may make them harder to read, write, and understand.
+
+For example, you could encode a decimal number as a `Long`, and then make Cedar functions to construct and compare decimals:
+```
+function mydecimal(i,f) {
+  if f >= 0 && f <= 9999 then
+    i * 10000 + f    // will overflow if i too big
+  else
+    18446744073709551615 + 1    // fraction f too big: induce overflow
+};
+function mydecimalLTE(d,e) {
+  d <= e    // d and e are just Long numbers
+}
+```
+These functions basically implement the equivalent of Cedar `decimal` numbers. But the approach has at least two drawbacks.
+
+First, if `i` and/or `f` are outside the allowed range, you will get a Cedar overflow exception at run-time. This exception is not as illuminating as the custom error emitted by the `decimal` extension function (whose message will be `"Too many digits"`).
+Moreover, custom errors from constructor parameter validity checks can be emitted during validation when using extension functions, but not when using Cedar functions.
+
+Second, there is no special parsing for Cedar functions. With Cedar's built-in `decimal`, you can write `decimal("123.12")` which more directly conveys the number being represented than does `mydecimal(123,12)`.
```

Comment:
Worse, `mydecimal(123,12)` actually means `mydecimal(123,0012)`

Reply:
Boom! Good catch.

Reply:
And FTR, `mydecimal(123,12)` represents the number 123.0012, which you didn't quite write above.